### PR TITLE
refactor: Java migration for namespaced features migration

### PR DESCRIPTION
**Pull request by jackgerrits**

### CODE REVIEW

File: java/src/main/c++/jni_spark_vw.cc [hunk 1]
```diff
@@ -384,24 +384,14 @@ JNIEXPORT void JNICALL Java_org_vowpalwabbit_spark_VowpalWabbitExample_clear(JNI
...
     JNIEnv* env, jobject exampleObj, jchar ns, jint weight_index_base, jdoubleArray values)
 {
   INIT_VARS
 
   try
   {
-    addNamespaceIfNotExists(all, ex, ns);
-
-    auto features = ex->feature_space.data() + ns;
+    auto& features = ex->feature_space.get_or_create_feature_group(weight_index_base, ns);
```
Comment (rajan-chari) [inline java/src/main/c++/jni_spark_vw.cc:394]: How does this work?  Just gets the first feature group that matches ns?

File: java/src/main/java/org/vowpalwabbit/spark/VowpalWabbitExample.java [hunk 2]
```diff
@@ -33,19 +33,34 @@
...
+
+        /**
+     * Adds the sparse features values to the supplied namespace.
+     *
+     * @param ns      the first character of the namespace.
+     * @param baseIndex the namespace hash (e.g. hash(namespace))
+     * @param indices the indices of each corresponding feature value.
+     * @param values  the feature values.
+     */
+    public native void addToNamespaceSparse(char ns, int baseIndex, int[] indices, double[] values);
```
Comment (rajan-chari) [inline java/src/main/java/org/vowpalwabbit/spark/VowpalWabbitExample.java:63]: Wondering if there should be another addToNamespace overload that takes the whole namespace string.

Comment (jackgerrits) [reply java/src/main/java/org/vowpalwabbit/spark/VowpalWabbitExample.java:63]: This is simply enabling the spark java interface to work with this. The spark java interface requires manual hashing of your own namespace and was an intentional decision when it was created for performance reasons.

File: java/src/main/c++/jni_spark_vw.cc (hunk 1 above)
Comment (jackgerrits) [reply java/src/main/c++/jni_spark_vw.cc:394]: To create a namespace you must know both the hash and index for it, so the creation function need the complete info for creation. Duplicates are not allowed, so if the hash already exists you get back that existing feature_group

**Pull request closed without merging**